MekHQ/src/mekhq/campaign/stratcon/StratconContractInitializer.java

```diff
@@ -0,0 +1,340 @@
+        //      the idea being to create a roughly rectangular playing field that,
+        //      if one deploys a scout lance each week to a different spot, can be more or less fully covered
+        
+        StratconTrackState retVal = new StratconTrackState();
+        retVal.setRequiredLanceCount(numLances);
+        
+        // set width and height
+        int numHexes = numLances * 28;
+        int height = (int) Math.floor(Math.sqrt(numHexes));
+        int width = numHexes / height;
+        retVal.setWidth(width);
+        retVal.setHeight(height);
+        
+        retVal.setScenarioOdds(scenarioOdds);
+        retVal.setDeploymentTime(deploymentTime);
+        
+        // todo: place terrain
+        
+        return retVal;
+    }
+    
+    /**
+     * Generates an array list representing the number of objects to place in a given number of tracks.
+     */
+    private static List<Integer> trackObjectDistribution(int numObjects, int numTracks) {
+        List<Integer> retVal = new ArrayList<>();
+        int leftOver = numObjects % numTracks;
+        
+        for (int track = 0; track < numTracks; track++) {
+            int trackObjects = numObjects / numTracks;
+            
+            // if we are unevenly distributed, add an extra one
+            if(leftOver > 0) {
+                trackObjects++;
+                leftOver--;
+            }
+            
+            retVal.add(trackObjects);
+        }
+        
+        // don't always front-load extra objects
+        Collections.shuffle(retVal);        
+        return retVal;
+    }
+    
+    /**
+     * Worker function that takes a trackstate and plops down the given number of facilities owned by the given faction
+     * Avoids places with existing facilities and scenarios, capable of taking facility sub set and setting strategic objective flag.
+     */
+    private static void initializeTrackFacilities(StratconTrackState trackState, int numFacilities, ForceAlignment owner, 
+            List<String> facilitySubset, boolean strategicObjective) {
+        for (int fCount = 0; fCount < numFacilities; fCount++) {
+            StratconFacility sf =
+                    facilitySubset == null ? 
+                    StratconFacilityFactory.getRandomFacility() :
+                    StratconFacilityFactory.getFacilityByName(facilitySubset.get(Compute.randomInt(facilitySubset.size())));
+            sf.setOwner(owner);
+            sf.setStrategicObjective(strategicObjective);
+            
+            int x = Compute.randomInt(trackState.getWidth());
+            int y = Compute.randomInt(trackState.getHeight());
+            StratconCoords coords = new StratconCoords(x, y);
+            
+            // make sure we don't put the facility down on top of anything else
+            while ((trackState.getFacility(coords) != null) ||
+                    (trackState.getScenario(coords) != null)) {
```

Comment:
Is it possible this will infinite loop?

Reply:
That would mean that the whole map is covered with facilities and scenarios. Not something that's going to happen normally, but it's probably possible to create some kind of edge case contract definition that results in this behavior.